### ISSUE TITLE
Having trouble testing POST to dev.to using httr

### ISSUE THREAD

**Issue by DaveParr** [ropensci/vcr#169]
I'm building an API package to the dev.to api [here](https://docs.dev.to/api/#operation/createArticle) and having trouble testing a post method.

My current version of the package is available [here](https://github.com/DaveParr/dev.to.ol), and the specific commit where the issue occurs is on dev branch [here](https://github.com/DaveParr/dev.to.ol/commit/e08d12c4cf0c59c113e922eb5d9e882571d4789a).

I'm sure I've missed something key, though I have been working through what I thought would be the relevant parts of the book, and specifically am trying to apply your [process for httr and testthat outlined here](https://books.ropensci.org/http-testing/gotchas.html).

What I am finding is that:

* when I `use_cassette` the first time manually I get a file that is populated as expected. That is also part of that commit, I put it in manually.
* when I run the test suite properly it fails, as the file that is created is actually blank, and so when read all properties are NULL

Is this possibly an authentication issue? I have up until now been using a .Renviron file (not on GH, for obv reasons). Can I rely on that to be found in the context of these tests? If not, how should I handle authentication in a case like this?

I also had the same behaviours when using mockr to stub it out. I think that was actually do to me not setting up the test environment correctly now and it was probably actually calling the real api. Am approaching that problem now, but it's very out of scope from this issue :)

```r
devtools::session_info()
─ Session info ───────────────────────────────────────────────────
 setting  value                       
 version  R version 3.6.0 (2019-04-26)
 os       macOS  10.15.4              
 system   x86_64, darwin15.6.0        
 ui       RStudio                     
 language (EN)                        
 collate  en_GB.UTF-8                 
 ctype    en_GB.UTF-8                 
 tz       Europe/London               
 date     2020-05-28                  

─ Packages ───────────────────────────────────────────────────────
 package     * version    date       lib source        
 assertthat    0.2.1      2019-03-21 [1] CRAN (R 3.6.0)
 attempt       0.3.1      2020-05-03 [1] CRAN (R 3.6.2)
 backports     1.1.6      2020-04-05 [1] CRAN (R 3.6.2)
 callr         3.4.3      2020-03-28 [1] CRAN (R 3.6.2)
 cli           2.0.2      2020-02-28 [1] CRAN (R 3.6.0)
 crayon        1.3.4      2017-09-16 [1] CRAN (R 3.6.0)
 crul          0.9.0      2019-11-06 [1] CRAN (R 3.6.0)
 curl          3.3        2019-01-10 [1] CRAN (R 3.6.0)
 desc          1.2.0      2018-05-01 [1] CRAN (R 3.6.0)
 dev.to.ol     0.0.0.9000 2020-05-28 [1] local         
 devtools      2.0.2      2019-04-08 [1] CRAN (R 3.6.0)
 digest        0.6.25     2020-02-23 [1] CRAN (R 3.6.0)
 ellipsis      0.3.0      2019-09-20 [1] CRAN (R 3.6.0)
 evaluate      0.14       2019-05-28 [1] CRAN (R 3.6.0)
 fansi         0.4.1      2020-01-08 [1] CRAN (R 3.6.0)
 fauxpas       0.5.0      2020-04-13 [1] CRAN (R 3.6.2)
 fs            1.3.1      2019-05-06 [1] CRAN (R 3.6.0)
 glue          1.4.0      2020-04-03 [1] CRAN (R 3.6.2)
 hms           0.5.3      2020-01-08 [1] CRAN (R 3.6.0)
 htmltools     0.3.6      2017-04-28 [1] CRAN (R 3.6.0)
 httpcode      0.3.0      2020-04-10 [1] CRAN (R 3.6.2)
 httr          1.4.1      2019-08-05 [1] CRAN (R 3.6.0)
 knitr         1.26       2019-11-12 [1] CRAN (R 3.6.0)
 lifecycle     0.2.0      2020-03-06 [1] CRAN (R 3.6.0)
 magrittr      1.5        2014-11-22 [1] CRAN (R 3.6.0)
 memoise       1.1.0      2017-04-21 [1] CRAN (R 3.6.0)
 packrat       0.5.0      2018-11-14 [1] CRAN (R 3.6.0)
 pillar        1.4.3      2019-12-20 [1] CRAN (R 3.6.0)
 pkgbuild      1.0.7      2020-04-25 [1] CRAN (R 3.6.2)
 pkgconfig     2.0.3      2019-09-22 [1] CRAN (R 3.6.0)
 pkgload       1.0.2      2018-10-29 [1] CRAN (R 3.6.0)
 prettyunits   1.1.1      2020-01-24 [1] CRAN (R 3.6.0)
 processx      3.4.2      2020-02-09 [1] CRAN (R 3.6.0)
 ps            1.3.2      2020-02-13 [1] CRAN (R 3.6.0)
 R6            2.4.1      2019-11-12 [1] CRAN (R 3.6.0)
 Rcpp          1.0.4      2020-03-17 [1] CRAN (R 3.6.0)
 readr         1.3.1      2018-12-21 [1] CRAN (R 3.6.0)
 remotes       2.0.4      2019-04-10 [1] CRAN (R 3.6.0)
 rlang         0.4.5      2020-03-01 [1] CRAN (R 3.6.0)
 rmarkdown     2.1        2020-01-20 [1] CRAN (R 3.6.0)
 rprojroot     1.3-2      2018-01-03 [1] CRAN (R 3.6.0)
 rstudioapi    0.11       2020-02-07 [1] CRAN (R 3.6.0)
 sessioninfo   1.1.1      2018-11-05 [1] CRAN (R 3.6.0)
 stringi       1.4.6      2020-02-17 [1] CRAN (R 3.6.0)
 stringr       1.4.0      2019-02-10 [1] CRAN (R 3.6.0)
 testthat      2.3.2      2020-03-02 [1] CRAN (R 3.6.0)
 tibble        3.0.1      2020-04-20 [1] CRAN (R 3.6.2)
 usethis       1.5.0      2019-04-07 [1] CRAN (R 3.6.0)
 vctrs         0.2.4      2020-03-10 [1] CRAN (R 3.6.0)
 webmockr      0.6.2      2020-03-24 [1] CRAN (R 3.6.0)
 whisker       0.3-2      2013-04-28 [1] CRAN (R 3.6.0)
 withr         2.2.0      2020-04-20 [1] CRAN (R 3.6.2)
 xfun          0.13       2020-04-13 [1] CRAN (R 3.6.2)

[1] /Library/Frameworks/R.framework/Versions/3.6/Resources/library
```

**Comment by DaveParr** [comment]
Just spotted chapter 13 on security. Going through that now. Sorry, I missed it earlier :)

**Comment by sckott** [comment]
Thanks for the issue @DaveParr ! 

Curious how you are running your tests. Did you see https://books.ropensci.org/http-testing/vcr-intro.html#check-vs-test

**Comment by DaveParr** [comment]
I did. I'm running them as tests AFAIK. Using cmd+shift+T or through the 'test' button at the top of my test file.

Thanks for the fast response :)

**Comment by sckott** [comment]
Okay, I see that cmd+shift+T runs `devtools::test()`. 

Hmm, cloned your repo down, looking at dev branch. Looks like you don't have all the pieces set up. Did you run `vcr::use_vcr()`? I'd run that in your repo - which adds vcr as a Suggest, adds a `helper-dev.to.ol.R` file in `tests/testthat`. After I did that I think it's working locally for me. 

Since there's auth keys here, definitely see https://books.ropensci.org/http-testing/security.html#api-keys and add `filter_sensitive_data` to your `vcr_configure()`  in the `helper-dev.to.ol.R` file

**Comment by DaveParr** [comment]
I did not use `use_vcr`, thanks. I'll try that out over the weekend :)

**Comment by DaveParr** [comment]
I see a little more I think now. I had very much misinterpreted how this worked. Very much my fault. A number of issues were occurring simultaneously, but I've now unpicked them.

So I have:

* realised now that the key actually was exposed in that commit referenced above, which I hadn't realised (my fault), and so I've revoked that key and regenerated it on dev.to. If I had noticed this it would have helped me with point 3.
* run `vcr::use_vcr` on the project, which I should have done first but didn't.
* set my helper to include `filter_sensitive_data = list("<<<my_api_key>>>" = Sys.getenv('DEVTO'))`. It took me a few passes to realise what was happening here. I had initially understood this to be a named list, where the name _was_ the api key. I don't know why I understood it like that, but I did for most of yesterday, which then lead me to.
* realised that my '.Rmd' file wasn't accessible to the tests when they were run with testthat due to their directory location. Moved them to the correct place.
* didn't have decent logic in the function about messaging/erroring that the file couldn't be found.

Now I have rectified all those, the test is working correctly. 

Thanks for putting both `webmockr` and `vcr` out and helping with this issue. I've never done much api-response-side-testing work in R, but I'm enjoying using them. It's easy when you know how :)
